Ignoring checks within a module

Hi All,

My first post here, I've looked to try and find an answer to my problem but couldn't see one, but if this is already resolved apologies but if you could link me to the fix that'd be great.  TFsec seems to be a great idea so thanks to everyone for their efforts.

I'm trying to use TFSec to review my code, including what is getting deployed into modules being used.

I've written the below piece of code, and successfully annotated the code to get TFSec to ignore a security group rule check, using the Hashicorp provided terraform module for security groups : -

```
#Create Security Group for Build Runner
module "build_runner_sg" {
  source = "terraform-aws-modules/security-group/aws"

  name        = "build_runner_sg"
  description = "Security group for Gitlab Build Runner"
  vpc_id      = module.vpc.vpc_id

  egress_with_cidr_blocks = [
    {
      from_port   = 0
      to_port     = 65535
      protocol    = -1
      description = "Runner outbound access"
      cidr_blocks = "0.0.0.0/0" #tfsec:ignore:AWS007 ignore warning as this open outbound security rule is valid
    }
  ]
}
```
Having done a Terraform init and then run tfsec I'm still getting TFSec warnings for the code in the module related to the above, e.g. : -
```
Problem 1
  [AWS007][WARNING] Resource 'module.build_runner_sg:aws_security_group_rule.egress_rules' defines a fully open egress security group rule.
  /builds/_redacted_/.terraform/modules/build_runner_sg/main.tf:440
     437 |   security_group_id = local.this_sg_id
     438 |   type              = "egress"
     439 | 
     440 |   cidr_blocks      = var.egress_cidr_blocks
     441 |   ipv6_cidr_blocks = var.egress_ipv6_cidr_blocks
     442 |   prefix_list_ids  = var.egress_prefix_list_ids
     443 |   description      = var.rules[var.egress_rules[count.index]][3]
  See https://tfsec.dev/docs/aws/AWS007/ for more information.
```
I can't find a way of ignoring this message from within the module, whilst still scanning the module.  I thought I could ignore checking modules (e.g. not running terraform init first), but it would be good to check everything if that's possible.

Mark

I have a similar issue that prevents me from using `tfsec` as a pre-commit check (I have to delete all `.terraform` directories before committing).

I am getting an error like:
```
[AWS002][ERROR] Resource 'module.a:module.b:module.c:module.s3:aws_s3_bucket.destination' does not have logging enabled.
```
...which is 4 levels deep from the top level project.

I thought I could simply suppress it at the top leve via:
```hcl
# tfsec:ignore:AWS002
module "a" {
...
```
...but alas that does not seem to have any effect.

It would be great if one could selectively apply suppressions to specific module instances, because in many cases the nested module cannot be easily modified.

--- a/example/moduleWithIgnore/main.tf
+++ b/example/moduleWithIgnore/main.tf
@@ -1,0 +1,26 @@
+#tfsec:ignore:AWS052:exp:2021-02-02
+module "db" {
+  source = "terraform-aws-modules/rds/aws"
+  version = "~> 2.0"
+
+  identifier = "demodb"
+
+  engine            = "mysql"
+  engine_version    = "5.7.19"
+  instance_class    = "db.t2.large"
+  allocated_storage = 5
+
+  name     = "demodb"
+  username = "user"
+  password = aws_ssm_parameter.pw.value
+  port     = "3306"
+
+  maintenance_window = "Mon:00:00-Mon:03:00"
+  backup_window      = "03:00-06:00"
+}
+
+resource "aws_ssm_parameter" "pw" {
+  name = "pw"
+  type = "SecureString"
+  value = "changeme"
+}

--- a/internal/app/tfsec/block/block.go
+++ b/internal/app/tfsec/block/block.go
@@ -34,6 +34,13 @@ func (block *Block) HasModuleBlock() bool {
 	return block.moduleBlock != nil
 }
 
+func (block *Block) GetModuleBlock() (*Block, error) {
+	if block.HasModuleBlock() {
+		return block.moduleBlock, nil
+	}
+	return nil, fmt.Errorf("the block does not have an associated module block")
+}
+
 func (block *Block) body() *hclsyntax.Body {
 	return block.hclBlock.Body.(*hclsyntax.Body)
 }

--- a/internal/app/tfsec/scanner/scanner.go
+++ b/internal/app/tfsec/scanner/scanner.go
@@ -68,7 +68,7 @@ func (scanner *Scanner) Scan(blocks []*block.Block) []result.Result {
 						results = append(results, *res)
 					} else if ruleResults != nil {
 						for _, ruleResult := range ruleResults.All() {
-							if scanner.includeIgnored || (!scanner.checkRangeIgnored(ruleResult.RuleID, ruleResult.Range, checkBlock.Range()) && !checkInList(ruleResult.RuleID, scanner.excludedRuleIDs)) {
+							if scanner.includeIgnored || (!scanner.checkRangeIgnored(ruleResult.RuleID, ruleResult.Range, checkBlock) && !checkInList(ruleResult.RuleID, scanner.excludedRuleIDs)) {
 								results = append(results, ruleResult)
 							} else {
 								// rule was ignored
@@ -84,18 +84,27 @@ func (scanner *Scanner) Scan(blocks []*block.Block) []result.Result {
 	return results
 }
 
-func (scanner *Scanner) checkRangeIgnored(id string, r block.Range, b block.Range) bool {
-	raw, err := ioutil.ReadFile(b.Filename)
+func readLines(filename string) ([]string, error) {
+	raw, err := ioutil.ReadFile(filename)
 	if err != nil {
-		return false
+		return nil, err
 	}
-	ignoreAll := "tfsec:ignore:*"
-	ignoreCode := fmt.Sprintf("tfsec:ignore:%s", id)
-	lines := append([]string{""}, strings.Split(string(raw), "\n")...)
+
+	return append([]string{""}, strings.Split(string(raw), "\n")...), nil
+}
+
+func (scanner *Scanner) checkRangeIgnored(id string, r block.Range, b *block.Block) bool {
+	lines, err := readLines(b.Range().Filename)
+	if err != nil {
+		debug.Log("the file containing the block could not be opened. %s", err.Error())
+	}
 	startLine := r.StartLine
 
+	ignoreAll := "tfsec:ignore:*"
+	ignoreCode := fmt.Sprintf("tfsec:ignore:%s", id)
+
 	foundValidIgnore := false
-	lineValidIgnoreFound := 0
+	var ignoreLine string
 
 	// include the line above the line if available
 	if r.StartLine-1 > 0 {
@@ -110,44 +119,71 @@ func (scanner *Scanner) checkRangeIgnored(id string, r block.Range, b block.Rang
 
 		if strings.Contains(lines[number], ignoreAll) || strings.Contains(lines[number], ignoreCode) {
 			foundValidIgnore = true
-			lineValidIgnoreFound = number
+			ignoreLine = lines[number]
 			break
 		}
 	}
 
 	// check the line above the actual resource block
-	if b.StartLine-1 > 0 {
-		line := lines[b.StartLine-1]
+	if b.Range().StartLine-1 > 0 {
+		line := lines[b.Range().StartLine-1]
 		if strings.Contains(line, ignoreAll) || strings.Contains(line, ignoreCode) {
 			foundValidIgnore = true
-			lineValidIgnoreFound = b.StartLine - 1
+			ignoreLine = line
 		}
 	}
 
-	if foundValidIgnore {
-		lineWithPotentialExp := lines[lineValidIgnoreFound]
-		expWithCode := fmt.Sprintf("%s:exp:", id)
-		if indexExpFound := strings.Index(lineWithPotentialExp, expWithCode); indexExpFound > 0 {
-			debug.Log("Expiration date found on ignore '%s'", lineWithPotentialExp)
-			layout := fmt.Sprintf("%s2006-01-02", expWithCode)
-			expDate := lineWithPotentialExp[indexExpFound : indexExpFound+len(layout)]
-			parsedDate, err := time.Parse(layout, expDate)
+	// if nothing found yet, walk up any module references
+	if !foundValidIgnore {
+		foundValidIgnore, ignoreLine = traverseModuleTree(b, ignoreAll, ignoreCode)
+	}
 
+	if foundValidIgnore {
+		expWithCode := fmt.Sprintf("%s:exp:", id)
+		if indexExpFound := strings.Index(ignoreLine, expWithCode); indexExpFound > 0 {
+			debug.Log("Expiration date found on ignore '%s'", ignoreLine)
+			layout := fmt.Sprintf("%s2006-01-02", expWithCode)
+			expDate := ignoreLine[indexExpFound : indexExpFound+len(layout)]
+			parsedDate, err := time.Parse(layout, expDate)
 			if err != nil {
 				// if we can't parse the date then we don't want to ignore the range
-				debug.Log("Unable to parse exp date in ignore: '%s'. The date format is invalid. Supported format 'exp:yyyy-mm-dd'.", lineWithPotentialExp)
+				debug.Log("Unable to parse exp date in ignore: '%s'. The date format is invalid. Supported format 'exp:yyyy-mm-dd'.", ignoreLine)
 				return false
 			}
-
 			currentTime := time.Now()
 			ignoreExpirationDateBreached := currentTime.After(parsedDate)
 			if ignoreExpirationDateBreached {
-				debug.Log("Ignore expired, check will be performed Filename: %s:%d", b.Filename, b.StartLine)
+				debug.Log("Ignore expired, check will be performed Filename: %s:%d", b.Range().Filename, b.Range().StartLine)
 			}
-
 			return !ignoreExpirationDateBreached
 		}
 	}
 
 	return foundValidIgnore
+}
+
+func traverseModuleTree(b *block.Block, ignoreAll, ignoreCode string) (bool, string) {
+
+	// check on the module
+	if b.HasModuleBlock() {
+		moduleBlock, err := b.GetModuleBlock()
+		if err != nil {
+			debug.Log("error occured trying to get the module block for [%s]. %s", b.FullName(), err.Error())
+			return false, ""
+		}
+		moduleLines, err := readLines(moduleBlock.Range().Filename)
+		if err != nil {
+			return false, ""
+		}
+		if moduleBlock.Range().StartLine-1 > 0 {
+			line := moduleLines[moduleBlock.Range().StartLine-1]
+			if strings.Contains(line, ignoreAll) || strings.Contains(line, ignoreCode) {
+				return true, line
+			}
+		}
+
+		return traverseModuleTree(moduleBlock, ignoreAll, ignoreCode)
+	}
+
+	return false, ""
 }

